File: src/stimy.pl
#!/usr/sbin/env /usr/sbin/perl
use PERLVERSION;
use strict;
use warnings;
no warnings 'uninitialized';
#use Data::Dumper;
my $sharp = '#';
my $asterisk = '\*';
my $slash = '/';
my $backslash = '\\';
my $singlequote = '\'';
my $doublequote = '"';
my $equal = '=';
my $empty = '';
my $log = undef;
my $sp = '[ \t\v\n\f]';
my $nonsp = '[^ \t\v\n\f]';
my $tab = '\t';
my $lbrace = '{';
my $rbrace = '}';
my $col = ':';
my $space = ' ';
my $semicol = ';';
my $nl = "\n";
my $indent = $space x 4;
my $ignoreword ='(?:__typeof__)';
my $assignmentop = '(?:=|\+=|\-=|\*=|/=|\%=|\<\<=|\>\>=|\&=|\^=|\|=)';
my $alpha = '[0-9a-zA-Z_\-]';
my $nonalpha = '[^0-9a-zA-Z_\-]';
my $wordsep = '(?:[ \t\n!\(\;=])';
my $anyword = '(?:[a-z][0-9a-zA-Z_\-]*)';
my $otherword = '(?!return|if|[a-zA-Z_][0-9a-zA-Z_\-]*)';
my $arguments='(?:[[:alnum:]]|[\_\,\%\\\&\-\(\>\.\*\"\:\[\]]|\s*)+';
my $lparent = '(';
my $rparent = ')';
my $rparentnosemicol = '\)(?!\;)';
my $begin = '^';
my $end = '\$';
# Left-parent index list.
my @pparent = ();
# Comment start and end position list.
my @pcomment = (0,0);
my %keyword = (
    error => 'error',
    pragma => 'pragma',
    operator => 'operator',
    elif => 'elif',
    line => 'line',
    endif => 'endif',
    ifdef => 'ifdef',
    include => 'include',
    undef => 'undef',
    defined => 'defined',
    auto => 'auto',
    char => 'char',   
    default => 'default',
    else => 'else',
    for => 'for',
    inline => 'inline',
    return => 'return',
    static => 'static',
    union => 'union',
    while => 'while',
    _Bool => '_Bool',
    _Complex => '_Complex',
    restrict => 'restrict',
    enum => 'enum',
    goto => 'goto',
    int => 'int',
    short => 'short',
    struct => 'struct',
    unsigned => 'unsigned',
    break => 'break',
    const => 'const',
    do => 'do',
    extern => 'extern',
    if => 'if',
    long => 'long',
    signed => 'signed',
    switch => 'switch',
    void => 'void',
    case => 'case',
    continue => 'continue',
    double => 'double',
    float => 'float',
    _Imaginary => '_Imaginary',
    register => 'register',
    sizeof => 'sizeof',
    typeof => 'typeof',
    typedef => 'typedef',
    volatile => 'volatile',
);
my %me = (
    replacedi => 0,
    rparenti => 0,
    headi => 0,
    taili => 0,
    preprocessor => 0,
    inputi => -1,
    comment => 0,
    dquote => 0,
    squote => 0,
    fundef => 0,
    lookahead => 0,
    pattern => ' ',
    nexti => 0,
    replacement => ' ',
    logfile => '/tmp/stimy.log',
    infile => $ARGV[0],
    i => 0,
    unicodesize => 256,
    num_brace => 0,
    preinput => "#ifndef STIMY_H\n#include <stimy.h>\n#endif\n",
    insertbegin => "${lbrace}${nl}${indent}stimy_pre()${semicol}",
    insertend => "${nl}${indent}stimy_post()${semicol}${nl}${rbrace}",
);
sub debug{
    say $log "$_[0]";
}
sub tab{
    ord($tab);
}
sub space{
    ord($space);
}
sub sharp{
    ord($sharp);
}
sub rparent{
    ord($rparent);
}
sub lparent{
    ord($lparent);
}
sub lbrace{
    ord($lbrace);
}
sub rbrace{
    ord($rbrace);
}
sub nl{
    ord($nl);
}
sub semicol{
    ord($semicol);
}
sub singlequote{
    ord($singlequote);
}
sub doublequote{
    ord($doublequote);
}
sub slash{
    ord($slash);
}
sub backslash{
    ord($backslash);
}
sub fnothing {
    return if($me{preprocessor} || $me{squote} || $me{dquote} || $me{comment});
    flookahead() if($me{lookahead});
}
sub fspace {;}
sub ftab {;}
# Decision table arguments: 0, ASCII character. 1, function pointer.  
sub hash {
    $me{unicode}[$_[0]] = $_[1];
}
sub fbackslash {
    debug("fbackslash:");
    $me{inputi}++;
}
sub fnl {
    return if($me{squote} || $me{dquote});
    debug("fnl:");
    # End of Comment format: '//'
    if($me{comment}){
        debug("end comment 2:");
        $me{comment} = 0; 
        # Append
        $pcomment[1] = $me{inputi};
    }
    if($me{preprocessor}){
        debug("end preprocessor.");
        $me{preprocessor} = 0;
    }
    hash(nl(),\&fnothing);
}
sub fsharp {
    return if($me{preprocessor} || $me{squote} || $me{dquote} || $me{comment});
    debug("fsharp: start preprocessor.");
    $me{preprocessor} = 1;
    hash(nl(),\&fnl);
}
sub fslash {
    return if($me{squote} || $me{dquote} || $me{preprocessor});
    # End of Comment format: '*/'.
    if($me{comment}){
        debug("fslash: end comment 1.");
        $_ = substr($me{input},$me{inputi} - 1,1);
        return if(!m;$asterisk;);
        $me{comment} = 0;
        # Append
        $pcomment[1] = $me{inputi};
        hash(nl(),\&fnl);
        return;
    }
    $_ = substr($me{input},$me{inputi} + 1,1);
    #  Start of the Comment format: '/*'
    return if(!m;$asterisk|$slash;);
    if(m;$asterisk;){
        debug("fslash: start comment 1.");
        hash(nl(),\&fnothing);
    }else{
        # Start of Comment format: '//'
        debug("fslash: start comment 2.");
        hash(nl(),\&fnl);
    }
    $me{comment} = 1;
    $me{inputi}++;
    # String between Comments.
    $_ = substr($me{input},$pcomment[1] + 1,$me{inputi} - $pcomment[1] - 2);
    return if(m;^$sp*$;);
    # Non consecutive Comments.
    $pcomment[0] = $me{inputi} - 1; 
}
sub fsinglequote {
    return if($me{dquote} || $me{comment} || $me{preprocessor});
    debug("fsinglequote:");
    # Flip the indicator.
    return $me{squote} = 0 if($me{squote});
    $me{squote} = 1;
}
sub fdoublequote {
    return if($me{squote} || $me{comment} || $me{preprocessor});
    debug("fdoublequote:");
    # Flip the indicator.
    return $me{dquote} = 0 if($me{dquote});
    $me{dquote} = 1;
}
sub flbrace {
    # Counting brace is only valid inside function definition.
    return if($me{squote} || $me{dquote} || $me{comment}
         || $me{preprocessor} || !$me{fundef});
    debug("flbrace:");
    return if(++$me{num_brace} > 1);
    $me{lbracei} = $me{inputi};
}
sub frbrace {
    return if($me{squote} || $me{dquote} || $me{comment}
         || $me{preprocessor} || !$me{lbracei});
    debug("frbrace:");
    return if(--$me{num_brace} >0);
    fheadtail();
}
sub fheadtail()
{
    debug("fheadtail:");
    debug("$me{lbracei}::$me{inputi}");
    $me{replacedi} = $me{lbracei}; 
    $me{replaced} = $lbrace;
    $me{replacement} = $me{insertbegin};
    freplace();
    $me{replacedi} = $me{inputi}; 
    $me{replaced} = $rbrace;
    $me{replacement} = $me{insertend};
    freplace();
    $me{fundef} = 0;
    $me{lbracei} = 0;
}
sub flparent {
    return if($me{squote} || $me{dquote} || $me{comment} || $me{preprocessor});
    debug("flparent:");
    return if($me{num_brace} < 1);
    push(@pparent,$me{inputi});
    debug("size:" . scalar(@pparent));
}
sub frparent {
    return if($me{squote} || $me{dquote} || $me{comment} || $me{preprocessor});
    debug("frparent:");
    # Possible function definition.
    return $me{fundef} = 1 if($me{num_brace} < 1);
    flookahead() if($me{lookahead});
    $me{rparenti} = $me{inputi};
    flookbehind();
    pop(@pparent);
    debug("size:" . scalar(@pparent));
}
sub flookbehind {
    $_ = substr($me{input},$me{inputi},1);
    debug("flookbehind:$_");
    if($pcomment[1] > $pparent[-1]){
        $me{i} = $pparent[-1] - 1;
        foneline();
        return;
    }
    $me{i} = $pcomment[1] + 1;
    $_ = substr($me{input},$me{i},$pparent[-1] - $me{i});
    if(m;$nonsp;){
        $me{i} = $pparent[-1] - 1; 
    }else{
        $me{i} = $pcomment[0] - 1; 
    }
    foneline();
}
sub foneline {
    debug("foneline:");
    $me{pattern} = $sp;
    for (;$me{i} > 0;$me{i}--){
        $_ = substr($me{input},$me{i},1);
        #debug(":$_:");
        next if(m;$me{pattern};);
        if(m;$sp;){
            $me{headi} = $me{i} + 1;
            last; 
        }
        $me{pattern} = $nonsp;
        $me{taili} = $me{i};
    }
    $_ = substr($me{input},$me{headi},$me{taili} - $me{i}); 
    debug("$me{headi}:$_:$me{taili}");
    return if(!m;^$anyword$;);
    return if($keyword{$_});
    $me{lookahead} = 1;
}
sub flookahead {
    $_ = substr($me{input},$me{inputi},1);
    debug("flookahead:$_");
    $me{lookahead} = 0;
    $_ = substr($me{input},$me{inputi},3);
    s{
        ^(
           =[^=] |
           [\+\-\*%\/\^\|]= |
           >>= | <<=
        )
    }{
        "$1" && return;
    }sex;
    $_ = substr($me{input},$me{headi},$me{taili} - $me{i});
    debug("$me{headi}:$_:$me{taili}");
    $me{replacedi} = $me{headi};
    $me{replaced} = substr($me{input},$me{headi},$me{rparenti} - $me{i});
    $me{replacement} = "stimy_echo($_,$me{replaced})";
    freplace();
}
sub freplace {
    debug("freplace:");
    $me{replaced_len} = length($me{replaced});
    substr($me{input},$me{replacedi},$me{replaced_len},$me{replacement});
    $me{increment} = length($me{replacement}) - $me{replaced_len}; 
    $me{input_len} += $me{increment};
    $me{inputi} += $me{increment};
#    debug("$me{headi}:$2:$me{rparenti}");
}
sub openfile {
    open($log, '>', "$me{logfile}") or die "Cann't open file: $me{logfile}. $!";
    open(INPUT, '<', "$me{infile}") or die "Cann't open file: $me{infile}. $!";
    local $/ = undef;
    $me{input} = <INPUT>;
    $me{input_len} = length($me{input});
}
sub run {
    # Prepare unicode Hash decision table.
    for (my $i = 0; $i < $me{unicodesize}; $i++){
        hash($i,\&fnothing);
    }
    # Prepare specific ASCII => function name decision.
    hash(space(),\&fspace);
    hash(tab(),\&ftab);
    hash(lparent(),\&flparent);
    hash(rparent(),\&frparent);
    hash(lbrace(),\&flbrace);
    hash(rbrace(),\&frbrace);
    hash(singlequote(),\&fsinglequote);
    hash(doublequote(),\&fdoublequote);   
    hash(slash(),\&fslash);
    hash(sharp(),\&fsharp);
    hash(backslash(),\&fbackslash);
    # Indirect function lookahead based on previous prepared hash table.
    while (++$me{inputi} < $me{input_len}){
        $me{unicode}[ord(substr($me{input},$me{inputi},1))]();
    }
}
sub postrun()
{
    debug("=====================");
    print "$me{preinput}$me{input}";
    close $log;
}
openfile();
run();
postrun();
#say Dumper(\%me);
#say Dumper(\@pcomment);

File: src/stimy.py
#!/bin/env -S PATH=/usr/local/bin:/usr/bin python3 -I
import re,tempfile,resource,glob,io,subprocess,sys
import os,socket,getpass,random,datetime
class Stimy:
    def __init__(self,*argv):
        self.message = {'-h':' print this help message.',
        '-t': ' test','-c': ' [source file]'}
        self.argv = argv
        self.args = argv[0]
        self.argc = len(self.args)
        if self.argc == 1: self.usage()
        self.option = { '-h':self.usage ,'-t':self.test,'-c':self.fdefault }
        self.uid = os.getuid()
        self.username = getpass.getuser()
        self.homedir = os.environ.get('HOME') + '/'
        self.tmpdir = '/var/tmp/'
        self.debugging = DEBUGGING
        self.content = ''
        self.tablesize = 256
        self.table = {}
        self.contenti = -1
        self.content_len = 0
        self.dquote = 0
        self.squote = 0
        self.preprocessor = 0
        self.backslash = 0
        self.log = '/tmp/stimy.log'
        self.fh = ''
        self.num_brace = 0
        self.fdef = 0
        self.idprocess = False
        self.idstarti = 0
        self.idendi = 0
        self.whitespace = {' ':1,'\t':1,'\n':1,'\r':1,'\f':1,'\v':1}
        self.indent = '    '
        self.semicol = ';'
        self.lowercase = 'abcdefghijklmnopqrstuvwxyz'
        self.uppercase = 'abcdefghijklmnopqrstuvwxyz'.upper()
        self.digits = '0123456789'
        self.underscore = '_'
        self.nl ='\n'
        self.parent = []
        self.fundefindicator = { ')':1, ';':1}
        self.preprocessorcondition = {'if':1,'ifdef':1,'ifndef':1}
        self.preprocessorend = {'endif':1}
        self.preprocessorstart = {'include':1,'define':1,'undef':1,'line':1,'error':1,'pragma':1}
        self.identifier = { 'a':1,'b':1,'c':1,'d':1,'e':1,'f':1,'g':1,'h':1,'i':1,'j':1,'k':1,'l':1,'m':1,'n':1,'o':1,'p':1,'q':1,'r':1,'s':1,'t':1,'u':1,'v':1,'w':1,'x':1,'y':1,'z':1,'A':1,'B':1,'C':1,'D':1,'E':1,'F':1,'G':1,'H':1,'I':1,'J':1,'K':1,'L':1,'M':1,'N':1,'O':1,'P':1,'Q':1,'R':1,'S':1,'T':1,'U':1,'V':1,'W':1,'X':1,'Y':1,'Z':1,'0':1,'1':1,'2':1,'3':1,'4':1,'5':1,'6':1,'7':1,'8':1,'9':1,'_':1}
        self.keyword = { 'error':1,'pragma':1,'operator':1,'elif':1,'line':1,'endif':1,'ifdef':1,'include':1,'undef':1,'defined':1,'auto':1,'char':1,'default':1,'else':1,'for':1,'inline':1,'return':1,'static':1,'union':1,'while':1,'_Bool':1,'_Complex':1,'restrict':1,'enum':1,'goto':1,'int':1,'short':1,'struct':1,'unsigned':1,'break':1,'const':1,'do':1,'extern':1,'if':1,'long':1,'signed':1,'switch':1,'void':1,'case':1,'continue':1,'double':1,'float':1,'_Imaginary':1,'register':1,'sizeof':1,'typeof':1,'typedef':1,'volatile':1}

    def comments(self):
       ## Remove multiline to one line.
        self.content = re.sub(r"\\\n",'',self.content)
        pattern = r"""
          //[^\n\r]*     ##  // ... comment on the same line
         |
                            ##  --------- COMMENT ---------
           /\*              ##  Start of /* ... */ comment
           [^*]*\*+         ##  Non-* followed by 1-or-more *'s
           (?:              ##
             [^/*][^*]*\*+  ##
           )*               ##  0-or-more things which don't start with /
                            ##    but do end with '*'
           /                ##  End of /* ... */ comment
         |                  ##  -OR-  various things which aren't comments:
           (                ##  Group 1 start
                            ##  ------ " ... " STRING ------
             "              ##  Start of " ... " string
             (?:            ##
               \\.          ##  Escaped char
             |              ##  -OR-
               [^"\\]       ##  Non "\ characters
             )*             ##
             "              ##  End of " ... " string
           |                ##  -OR-
                            ##
                            ##  ------ ' ... ' STRING ------
             '              ##  Start of ' ... ' string
             (?:            ##
               \\.          ##  Escaped char
             |              ##  -OR-
               [^'\\]       ##  Non '\ characters
             )*             ##
             '              ##  End of ' ... ' string
           |                ##  -OR-
                            ##  ------ ANYTHING ELSE -------
             .              ##  Anything other char
             [^/"'\\]*      ##  Chars which doesn't start a comment, string or escape
           )                ##  Group 1 end  
        """
        regex = re.compile(pattern,re.VERBOSE|re.MULTILINE|re.DOTALL)
        content = ''
        for m in regex.finditer(self.content):
            if m.group(1):content += m.group(1)
        self.content = re.sub(r'\s*\n','\n',content)
        self.content = re.sub(r'^\s*\n\s*','',self.content,1)
        self.content = re.sub(r'\s*\n\s*$','',self.content,1)

    def fdefault(self):
        if self.argc < 3: self.usage(self.args[1])
        if self.argc >= 2: self.sourcefile = self.args[2]
        self.fh = open(self.log,'a')
        with open(self.sourcefile,'r') as fh:
            self.content = fh.read()
        self.comments()
        # Build decision table for each content char. 
        for i in self.identifier:
            self.table[ord(i)] = self.fidentifier
        self.table[ord(' ')] = self.fspace 
        self.table[ord('\t')] = self.ftab 
        self.table[ord('\n')] = self.fnl 
        self.table[ord('#')] = self.fsharp 
        self.table[ord('/')] = self.fslash 
        self.table[ord('\\')] = self.fbackslash 
        self.table[ord('\'')] = self.fsinglequote 
        self.table[ord('\"')] = self.fdoublequote
        self.table[ord('(')] = self.flparent 
        self.table[ord(')')] = self.frparent
        self.table[ord('{')] = self.flbrace 
        self.table[ord('}')] = self.frbrace
        for i in range(self.tablesize):
            if i not in self.table: self.table[i] = self.fnothing
        self.content = list(self.content)
        self.content_len = len(self.content)
        while self.contenti < self.content_len: 
            self.table[ord(self.content[self.contenti][0])]()
            self.contenti += 1
        self.flog('=======================')
        print('#ifndef STIMY_H\n#include <stimy.h>\n#endif\n' + ''.join(self.content))

       # First condition to handle 
    def fsinglequote(self):
        if self.dquote or self.preprocessor:return
        if self.backslash: 
            self.backslash = 0
            return
        if self.squote:
            self.flog('fsinglequote:end')
            self.squote = 0
            return
        self.flog('fsinglequote:start')
        self.squote = 1

    def fdoublequote(self):
        if self.squote or self.preprocessor:return
        if self.backslash: 
            self.backslash = 0
            return
        if self.dquote:
           self.flog('fdoublequote:end')
           self.dquote = 0  
           return
        self.flog('fdoublequote:start')
        self.dquote = 1

    def fsharp(self):
        if self.dquote or self.squote:return
        # Lookforward find Preprocessor Keyword start index
        endi = self.content_len
        starti = self.contenti+1
        if starti >= self.content_len:return
        for i in range(starti,self.content_len):
            if self.content[i] == ' ':continue 
            starti = i
            break
        if self.content[starti] not in self.identifier:return
        # Lookforward find Preprocessor Keyword End Index
        for i in range(starti,self.content_len):
            if self.content[i] in self.identifier:continue
            endi = i
            break
        # Found Keyword
        word = ''.join(self.content[starti:endi]) 
        if not self.preprocessor and word in self.preprocessorstart:
            self.flog('fsharp:start preprocessor')
            self.preprocessor = 1
            return
        if not self.preprocessor and word in self.preprocessorcondition:
            self.flog('fsharp:start preprocessor condition')
            self.preprocessor = 2
            return
        if self.preprocessor and word in self.preprocessorend:
            self.flog('fsharp:end preprocessor')
            self.preprocessor = 0
            return

    def fnl(self):
        if self.dquote or self.squote:return
        if self.preprocessor == 1:
            self.flog('fnl:end preprocessor')
            self.preprocessor = 0
            return
#        self.flog('fnl:')

    def fbackslash(self):
        if self.backslash: 
            self.flog('fbackslash:end')
            self.backslash = 0
            return
        # Look forward one char
        if '\'\"'.find(self.content[self.contenti+1]) < 0:return
        self.flog('fbackslash:start')
        self.backslash = 1

    def flbrace(self):
        if self.dquote or self.squote or self.preprocessor:return
        # Increment Brace only inside function def
        if self.fdef:
            self.flog('flbrace:start')
            self.num_brace += 1
            return
        # Lookbehind Untill NonWithSpace Char
        for i in range(self.contenti-1,0,-1): 
            if self.content[i] in self.whitespace:continue
            if self.content[i] not in self.fundefindicator:return
            break
        # Found Function Definition
        self.flog('flbrace:fdef start')
        self.fdef = 1
        self.content[self.contenti] += self.nl + self.indent + 'stimy_pre()' + self.semicol 

    def frbrace(self):
        # Increment Brace only inside function def
        if self.dquote or self.squote or self.preprocessor or not self.fdef:return
        if self.num_brace > 0:
            self.flog('frbrace:end')
            self.num_brace -= 1
            return
        self.flog('frbrace:fdef end')
        self.fdef = 0
        self.content[self.contenti-1]+=self.indent + 'stimy_post()' + self.semicol + self.nl 

    def fidentifier(self):
        if self.dquote or self.squote or self.preprocessor or not self.fdef:return
        # Lookforward one char
        if self.content[self.contenti + 1][0] in self.identifier:
            if self.idprocess:return
            # First but not the only char in Identifier
            self.idstarti = self.contenti
            self.idprocess = True
            return

        # End of Identifier
        if not self.idprocess: self.idstarti = self.contenti
        self.idprocess = False
        if self.content[self.idstarti].isdigit():return
        self.idendi = self.contenti + 1
        identifier = ''.join(self.content[self.idstarti:self.idendi])
        if identifier == 'return':self.freturn()
        self.flog('fiden:' + identifier)

    # Found Return statements
    def freturn(self):
        for i in range(self.idendi,self.content_len):
            if self.content[i].find(';') < 0:continue
            self.content[self.idstarti-1] += 'stimy_post(' 
            self.content[i] += ');'
            return

    def flparent(self):
        if self.dquote or self.squote or self.preprocessor or not self.fdef:return
        self.flog('flparent:start')
        self.parent.append(self.contenti)

    def frparent(self):
        if self.dquote or self.squote or self.preprocessor or not self.fdef:return
        self.flog('frparent:end1')
        if len(self.parent) < 1:return
        lparenti = self.parent.pop()
        idstarti = idendi = 0
        semicolon = ''
        semicoloni = -1
        # Lookbehind find the End of Identidier
        for i in range(lparenti-1,0,-1):
            if self.content[i][0] in self.whitespace:continue
            if self.content[i][0] in self.identifier:
                idendi = i + 1
                break
            return
        # Lookbehind find the Start of Identidier
        for i in range(idendi-1,0,-1):
            if self.content[i][0] in self.identifier:continue
            if self.content[i][0] in self.whitespace:
                idstarti = i + 1
                break
            return

        word = ''.join(self.content[idstarti:idendi]).strip()
        self.flog('frparent:end2 ' + str(idstarti) + ' ' + str(idendi) + ' ' + word)
        # Ignore Keywords
        if word in self.keyword:return
        # Ignore Preprocessor
        if word.isupper():return
        # Ignore internal Identifier
        if word.startswith('_'):return

        self.content[idstarti-1] += ' stimy_echo(' + word + ','
        self.content[self.contenti] += ')'

    def flog(self,info=''):
        if self.debugging:print(info,file=self.fh)

    def fnothing(self):
        return
    def fspace(self):
        return
    def ftab(self):
        return
    def fslash(self):
        return
    def fundef(self):
        return

    def ftable(number,fname):
        self.table[number] = fname
        
    def test(self):
        with tempfile.NamedTemporaryFile(mode='w+',
        dir=self.tmpdir,delete=False) as self.testfh:
            self.testfh.write('big')
        proc = self.run(cmd='cat',stdout=subprocess.PIPE,infile=self.testfh.name)
        if proc != None:print(proc.stdout)
        proc = self.run(cmd='cat',stdout=subprocess.PIPE,infile='/etc/hostname')
        if proc != None:print(proc.stdout)
        self.run(cmd='date -u')

    def run(self, cmd='',infile='',outfile='',stdin=None,stdout=None,
        text=True,pass_fds=[],exit_errorcode='',shell=False):
        try:
            proc = None
            emit = __file__ + ':' + sys._getframe(1).f_code.co_name + ':' \
            + str(sys._getframe(1).f_lineno)
            if infile != '': stdin = open(infile,'r')
            if outfile != '': stdout = open(outfile,'w')
            proc = subprocess.run(cmd.split(),
            stdin=stdin,stdout=stdout,text=text,check=True,
            pass_fds=pass_fds,shell=shell)
            if infile != '': stdin.close() 
            if outfile != '': stdout.close()
            if not isinstance(proc,subprocess.CompletedProcess):
                self.debug(info='end 1',emit=emit)
                return None
            if isinstance(proc.stdout,str):
                proc.stdout = proc.stdout.rstrip('\n')
                self.debug(info='end 2',emit=emit)
                return proc
        except subprocess.CalledProcessError as e:
            emit += ':' + str(e.returncode)
            if exit_errorcode == '':
                if e.returncode != 0:
                    self.debug(info='end 3: ',emit=emit)
                    exit(1)
            elif e.returncode == exit_errorcode:
                self.debug(info='end 4',emit=emit)
                exit(1)
            return None
    
    def usage(self,option=1):
        if option in self.message:
            print(self.message[option].replace("@","\n    "))
        else:
            for key in self.message:
                print(key,self.message[key].replace("@","\n    "))
        exit(1)
    def debug(self,info='',outfile='',emit=''):
        if not self.debugging: return
        emit = sys._getframe(1).f_code.co_name + ':' \
        + str(sys._getframe(1).f_lineno) + ':' + info + ':' + emit
        print(emit)

if __name__ == '__main__':
    stimy = Stimy(sys.argv)
    if stimy.args[1] not in stimy.option: stimy.usage()
    try:
        stimy.option[stimy.args[1]]()
    except KeyboardInterrupt:
        stimy.debug(info='user ctrl-C')
    finally:
        stimy.debug(info='session finally end')
        for key,value in stimy.__dict__.items():
            if isinstance(value,io.TextIOWrapper):
                value.close()
                continue
            if isinstance(value,tempfile._TemporaryFileWrapper):
                value.close() 
                if os.access(value.name,os.R_OK): os.unlink(value.name)
        if stimy.debugging:
            with open('/tmp/stimylog','w') as stimy.logfh:
                print(stimy.__dict__,file=stimy.logfh)
